### ISSUE TITLE
[RFC] Product and taxon permalinks + slugs

### ISSUE THREAD

**Issue by pjedrzejewski** [Sylius/Sylius#3977]
Another round of brainstorm! Currently our handling of permalinks and slugs is a bit broken. Ideally we have all this editable in the backend etc. but this takes few things to resolve first.

I'd like to propose the following approach as the first step to get it handled properly:
- ProductTranslation
  - Has `name` property;
  - Has `slug` property, which is **always** slugified name, no matter what;
  - Has `permalink` property, which is generated by `ProductPermalinkGeneratorInterface` implementation;
- TaxonTranslation
  - Has `name` property;
  - Has `slug` property, which is **always** slugified name, no matter what;
  - Has `permalink` property, which is generated by `TaxonPermalinkGeneratorInterface` implementation;

We could provide different generator implementations by default, for example:
- One that simply uses "slug"
- One that uses "locale/slug" or "locale-slug", etc.
- One that uses "mainTaxon permalink/product slug", etc.
- One that does not generate any permalink and allows user to enter it manually
- And of course you could implement custom

Routing to display products / taxons would look:

``` yaml
sylius_shop_product_show:
    path: /products/{permalink}
    defaults:
        _controller: sylius.controller.product:showAction
        _sylius:
            repository:
                method: findOneByPermalink
                arguments: [$permalink]

sylius_shop_product_index_by_taxon:
    path: /{permalink}
    defaults:
        _controller: sylius.controller.product:indexAction
        _sylius:
            grid: sylius_shop_product_by_taxon
```

Grid would take care of filtering products by taxon with given permalink, etc. And would use Doctrine/ORM or ElasticSearch driver. :D

Things to figure out:
- How to keep permalinks sync? First thought would be a Doctrine listener but I'd like to avoid that whenever possible. I think we could process permalink in the form. That would also allow us to have something like "lock permalink" checkbox to avoid regenerating it, etc.

Of course there is much more to handle, like rewrites and so on but first things first.

Let me hear your thoughts!

**Comment by peteward** [comment]
The locking of permalinks is very important for SEO - in most cases you really don't want indexed product URLs to change!

Our product URLs (permalinks) are composed of many attributes relating to the product:

```
/p/highneck-coat-womens-melania-in-black/
```

We intend to use the Metadata PR to build product URLs based on a formula - in this case:

```
{{ archetype.name }}-{{ product.gender }}-{{ product.slug }}-in-{{ product.colour.name }}
```

However the permalink generator cannot be run against a product unless all attributes exist to fulfil this formula. There also needs to be a default `permalink` though, so we will start with setting the product's `permalink` to it's `code` then whenever a product is saved we'll check (probably in a lifecycle listener) if `permalink` == `code` and apply the permalink generator to try to replace it with the formulated version.

This way we only set the permalink once (automatically anyway) although we'll allow admin users to manually override if they need to.

That's a rough plan anyway!

**Comment by gabiudrescu** [comment]
I was also about to open a RFC called "Immutable slugs" but it seems Paweł was one step ahead of me.

The issue we are currently trying to address is exactly the one brought into discussion here: how do you make sure you generate a slug to use in the URL that remains unchanged after product is being indexed by search engines or its being shared over social networks?

Same for categories and any resource accessible by an URL (filters, attributes, custom entities etc.), not only products.

We were also taking into consideration an option of keeping track of all these changes and redirect with 301 from old to new.

We have a link like http://localhost:8000/p/cittizen-watch-bm6935-53a-gents-inglish. As  you can see, there's a typo in the brand name, Citizen having an extra T and a misspell for locale.

What do you think about the idea of keeping track of changes? Do you plan to tackle it in anyway or not?

Also, another thing I'd like to ask: why is there a permalink and a slug? why both? what's the purpose of them?

**Comment by peteward** [comment]
@gabiudrescu `slug` is (usually) the url-formatted representation of that entity's name, where as the `permalink` is (again usually) the full path to where the resource can be accessed on the site. It could include parent `slug` components if you are in a Taxonomy hierarchy.

E.g. if you have Grandparent taxon with slug `mens` then a Parent with slug `sale` and a Child with slug `t-shirts` the permalink on the child might be `mens/sale/t-shirts`

**Comment by pjedrzejewski** [comment]
Fixed in #6597.
